### PR TITLE
Upgrade to cosmos-sdk v0.45.0

### LINKED ISSUE #501: Update to Cosmos SDK v0.44
Currently, pre-release 
https://github.com/cosmos/cosmos-sdk/releases/tag/v0.43.0-alpha1

may be worth basing 1.0 RFC off here, as it contains quite a few breaking changes and the future state sync hook patch will be on top of it

### PR DESCRIPTION
**Pull request by alpe**
Resolves #501 

This upgrade includes
* Upgrade to cosmos-sdk version [v0.45.0](https://github.com/cosmos/cosmos-sdk/releases/tag/v0.45.0)
* Upgrade to new IBC version: [v2.0.2](https://github.com/cosmos/ibc-go/releases/tag/v2.0.2)
* Use **32 byte** address length for contract addresses
* Replace msg.Route() based handlers with ADR 031 request type routing for contract messages
* Add [feegrant](https://github.com/cosmos/cosmos-sdk/tree/v0.44.5/x/feegrant) module to demo app
* Add [authz](https://github.com/cosmos/cosmos-sdk/tree/v0.44.5/x/authz) module to demo app

Follow ups:
- Should we add support for "feegrant" to cosmwasm?
- Should we add support for "authz" to cosmwasm?
- Should we integrate IBC version negotiation, https://github.com/CosmWasm/wasmvm/issues/269?
- Should we pass the IBC relayer address to the contract?
- Should we do the IBC receive package (n)ack error cases differently?

### CODE REVIEW

**Comment by ethanfrey** [comment]
Thank you @alpe for this PR.

Re the follow ups. This is a very important question before merging, or at least before tagging a release.

* Should we do the IBC receive package (n)ack error cases differently?

-> This is an important question we should work out in the design docs and make a decision there before implementing. This has been heavily discussed in https://github.com/CosmWasm/wasmd/issues/697 and changes were made to the wasmvm interface to expose all needed info to the runtime, but we need to document the behavior.

These ones would be nice to have by 1.0 but require extending the CosmWasm interfaces in a backwards compatible way:

* Should we add support for "feegrant" to cosmwasm?
* Should we add support for "authz" to cosmwasm?
* Should we integrate IBC version negotiation, Support IBC version negotiation in contracts wasmvm#269?
* Should we pass the IBC relayer address to the contract?

-> I added a tracking issue for this in CosmWasm https://github.com/CosmWasm/cosmwasm/issues/1200 Please add more thoughts/comments on these to that issue

**Comment by ethanfrey** [comment]
@faddat @iramiller @ValarDragon @RiccardoM @litvintech

I would love any comments here as you all have been working on 0.44 forks.

**Comment by channa-figure** [comment]
I really like the PR the changes are clean and look good.  I have been looking forward to seeing the differences and similarities between our forks changes.

However, we were hoping that issue #696 would have made it in this PR.  I see it is currently target for milestone v1.0.0.  Is it possible to bump it into this?  If not, we will just maintain a fork and wait until v1.0.0.

**Comment by faddat** [comment]
Hi!

a moment to collect thoughts :)

**Comment by ethanfrey** [comment]
All of those require extensions to the CosmWasm interface.
I would get this in with the same API, and then work on those once we support them in CosmWasm (another release).
There is a linked tracking issue in CosmWasm to discuss this as well.

**Comment by ethanfrey** [comment]
> > Should we pass the IBC relayer address to the contract?
> I'm a reasonably large relayer, and I'm not sure what benefit that could bring -- in fact, it may be best to keep contracts unaware of who is doing the relaying since we (relayers) provide a generic service.

This will be needed if we want to provide bonuses / fees / incentives to relayers in the future. It should not be used for auth

**Comment by ethanfrey** [comment]
Unless anyone has concerns, I feel this is good to merge

**Review by ethanfrey — APPROVED** [review]
good stuff

File: go.mod [hunk 1]
```diff
@@ -101,24 +108,30 @@ require (
...
+	gopkg.in/ini.v1 v1.66.2 // indirect
 	gopkg.in/yaml.v3 v3.0.0-20210107192922-496545a6307b // indirect
+	nhooyr.io/websocket v1.8.6 // indirect
 )
 
 replace (
+	github.com/99designs/keyring => github.com/cosmos/keyring v1.1.7-0.20210622111912-ef00f8ac3d76
 	github.com/confio/ics23/go => github.com/confio/ics23/go v0.6.6
-	github.com/cosmos/cosmos-sdk => github.com/cosmos/cosmos-sdk v0.42.11
+	github.com/cosmos/cosmos-sdk => github.com/cosmos/cosmos-sdk v0.45.0-rc1
```
Comment (tomtau) [inline go.mod:126]:
```suggestion
	github.com/cosmos/cosmos-sdk => github.com/cosmos/cosmos-sdk v0.45.0
```

Comment (ethanfrey) [reply go.mod:126]: @alpe why is this even here?
`github.com/cosmos/cosmos-sdk v0.45.0` is set above and we don't change repos.

I would remove it

**Review by ethanfrey — COMMENTED** [review]
good point

File: go.mod (hunk 1 above)
Comment (tomtau) [reply go.mod:126]: I think it's because of ibc-go which may be bringing in a different version of cosmos-sdk

Comment (ethanfrey) [reply go.mod]: Turns out they were added here: https://github.com/CosmWasm/wasmd/pull/692/files

With a recent comment they are not a good idea: https://github.com/CosmWasm/wasmd/pull/692#issuecomment-1017078882

I fixed them with your hint, but maybe they need to be removed as well....

Comment (alpe) [reply go.mod]: ja, a transitive dependency. I did not look where it came from but just wanted to ensure we have the right sdk version set.

**Comment by alpe** [comment]
lgtm 🎉

**Pull request merged**